File: Top_Interview_150/Hashmap/python/205.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:

        if not sorted(s.count(c) for c in set(s)) == sorted(t.count(c) for c in set(t)):
            return False

        n = len(s)
        dct = {c: c for c in set(s)}

        for i in range(n):
            if s[i] != t[i]:
                dct[s[i]] = t[i]

        new_s = "".join(dct[c] for c in s)
        print(new_s)

        return new_s == t


solution = Solution()
print(solution.isIsomorphic(s="badc", t="baba"))

File: Top_Interview_150/Hashmap/cpp/205.cpp
#include <vector>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution
{
public:
    bool isIsomorphic(string s, string t)
    {
        unordered_set<char> set_s(s.begin(), s.end());
        unordered_set<char> set_t(t.begin(), t.end());

        int n = s.size();

        vector<int> v1;
        vector<int> v2;
        for (char c : set_s)
            v1.push_back(count(s.begin(), s.end(), c));

        for (char c : set_t)
            v2.push_back(count(t.begin(), t.end(), c));

        sort(v1.begin(), v1.end());
        sort(v2.begin(), v2.end());

        if (v1 != v2)
            return false;

        unordered_map<char, char> dct;
        for (char c : set_s)
            dct[c] = c;

        for (int i = 0; i < n; ++i)
            if (s[i] != t[i])
                dct[s[i]] = t[i];

        string new_s;
        for (char c : s)
            new_s += dct[c];

        return new_s == t;
    }
};